### PR TITLE
Nano: Add tensorflow quantization example.

### PR DESCRIPTION
**Pull request by mycprotein**
### 1. Why the change?

Add an example which shows the quantization of model during inference.

### CODE REVIEW

File: python/nano/tutorial/inference/tensorflow/mnist.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+# pip install 'neural-compressor'
+
+# calib_dataset only accept tf.data.Dataset object
+tune_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+
+# Execute quantization
+q_model = model.quantize(precision='int8',
+                         accelerator=None,
+                         calib_dataset=tune_dataset,
+                         accuracy_criterion={
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/mnist.py:106]: we would like to present the easiest API calling to our users in fundamental example, so we may leave parameters after `accuracy_criterion` to be default values.

File: python/nano/tutorial/inference/tensorflow/mnist.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                         )
+
+y_test_hat = q_model(x_test)
+loss = float(tf.reduce_mean(
+             losses.categorical_crossentropy(y_test, y_test_hat)))
+categorical_accuracy = metrics.CategoricalAccuracy()
+categorical_accuracy.update_state(y_test, y_test_hat)
+accuracy = categorical_accuracy.result().numpy()
+print("Quantization test loss:", loss)
+print("Quantization test accuracy:", accuracy)
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/mnist.py:121]: how is the accuracy drop?

File: python/nano/tutorial/inference/tensorflow/mnist.py (hunk 1 above)
Comment (mycprotein) [reply python/nano/tutorial/inference/tensorflow/mnist.py:106]: OK, I will try to simplify it.

File: python/nano/tutorial/inference/tensorflow/mnist.py (hunk 2 above)
Comment (mycprotein) [reply python/nano/tutorial/inference/tensorflow/mnist.py:121]: I will add it to a comment.

File: python/nano/tutorial/inference/tensorflow/mnist.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+# Raw model test loss: 0.024767747148871422
+# Raw model test accuracy: 0.9918000102043152
+# Quantized model test loss: 0.02494174614548683
+# Quantized model test accuracy: 0.9917
+# Accuracy loss: 0.01%
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/mnist.py:117]: that's reasonable accuracy drop. But we may just use some qualitative way to describe this in comment (e.g. accuracy drop is with on 0.1%). That's because the quantization may bring different result every time you run it.

File: python/nano/tutorial/inference/tensorflow/mnist.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+# Quantization using Intel Neural Compressor
+# pip install 'neural-compressor'
+
+# calib_dataset only accept tf.data.Dataset object
+tune_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+
+# Execute quantization
+q_model = model.quantize(calib_dataset=tune_dataset)
+
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/mnist.py:104]: add comment "Inference using quantized model"

File: python/nano/tutorial/inference/tensorflow/mnist.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+y_test_hat = q_model(x_test)
+loss = float(tf.reduce_mean(
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/mnist.py:106]: add comment "evaluate the quantized model" and make some empty lines here

**Review by TheaperDeng — COMMENTED** [review]
others LGTM

File: python/nano/tutorial/inference/tensorflow/mnist.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+categorical_accuracy = metrics.CategoricalAccuracy()
+categorical_accuracy.update_state(y_test, y_test_hat)
+accuracy = categorical_accuracy.result().numpy()
+
+print("Quantization test loss:", loss)
+print("Quantization test accuracy:", accuracy)
+# Raw model test loss: 0.024767747148871422
+# Raw model test accuracy: 0.9918000102043152
+# Quantized model test loss: 0.02494174614548683
+# Quantized model test accuracy: 0.9917
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/mnist.py:122]: remove these specific evaluation number

**Comment by jason-dai** [comment]
MNIST is too trivial.

We should used some other (simple but more meaningful) model/data.

**Comment by mycprotein** [comment]
I will change the dataset to CIFAR10, and use ResNet as model. Maybe they are more meaningful and modern.

**Comment by mycprotein** [comment]
Dataset is CIFAR10 now, but the model is still a toy model. I have tried some relatively large models and mobilenet v2, they all take too long to train, so I think a toy model is better here.

File: python/nano/tutorial/inference/tensorflow/cifar10.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+model.add(layers.Conv2D(128, (3,3), padding='same', activation='relu'))
+model.add(layers.BatchNormalization())
+model.add(layers.MaxPooling2D(pool_size=(2,2)))
+model.add(layers.Dropout(0.5))
+
+model.add(layers.Flatten())
+model.add(layers.Dense(128, activation='relu'))
+model.add(layers.BatchNormalization())
+model.add(layers.Dropout(0.5))
+model.add(layers.Dense(num_classes, activation='softmax'))
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/cifar10.py:90]: put these in a `create_model` function

File: python/nano/tutorial/inference/tensorflow/cifar10.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+model.add(layers.Flatten())
+model.add(layers.Dense(128, activation='relu'))
+model.add(layers.BatchNormalization())
+model.add(layers.Dropout(0.5))
+model.add(layers.Dense(num_classes, activation='softmax'))  
+
+# this line is optional
+# model = Model(inputs=model.inputs, outputs=model.outputs)
+
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/cifar10.py:94]: put following part under `if __name__ == "__main__":`

File: python/nano/tutorial/inference/tensorflow/cifar10.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+             losses.categorical_crossentropy(y_test, y_test_hat)))
+categorical_accuracy = metrics.CategoricalAccuracy()
+categorical_accuracy.update_state(y_test, y_test_hat)
+accuracy = categorical_accuracy.result().numpy()
+
+print("Quantization test loss:", loss)
+print("Quantization test accuracy:", accuracy)
+# Note: accuracy decrease varies from different tasks and situations,
+# and sometimes the accuracy is even higher.
+# You can set a quantization threshold when making a quantization model.
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/cifar10.py:134]: how long does this example run?

Comment (mycprotein) [reply python/nano/tutorial/inference/tensorflow/cifar10.py:134]: About half an hour.

Comment (TheaperDeng) [reply python/nano/tutorial/inference/tensorflow/cifar10.py]: a little bit long for a fundamental example yah? 

@MeouSker77 how much time does it take to run through https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/training/tensorflow/tensorflow_train_multi_instance.py

I wonder if we can use this model and data in this example.

No comment on other parts of the code

Comment (MeouSker77) [reply python/nano/tutorial/inference/tensorflow/cifar10.py]: > @MeouSker77 how much time does it take to run through https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/training/tensorflow/tensorflow_train_multi_instance.py

about 7 minutes

Comment (MeouSker77) [reply python/nano/tutorial/inference/tensorflow/cifar10.py]: > I wonder if we can use this model and data in this example.

I remember using the model and data in that example may encounter the conflict of tcmalloc and tf quantization

Comment (MeouSker77) [reply python/nano/tutorial/inference/tensorflow/cifar10.py]: > > @MeouSker77 how much time does it take to run through https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/training/tensorflow/tensorflow_train_multi_instance.py
> 
> about 7 minutes

I uses `num_epochs = int(os.environ.get('NUM_EPOCHS', 10))` to get the epoch num, then `export NUM_EPOCHS=1` in bash script before running that example to accelerate it.

Comment (mycprotein) [reply python/nano/tutorial/inference/tensorflow/cifar10.py]: In my machine, this example takes about an hour to finish. (10 epochs)
The cifar10 example takes 1.5 minutes for an epoch, and half an hour in total. (20 epochs)

File: python/nano/tutorial/inference/tensorflow/cifar10.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This file is adapted from Keras Tutorial.
+# https://github.com/keras-team/keras-io/blob/
+# master/examples/vision/mnist_convnet.py
```
Comment (TheaperDeng) [inline python/nano/tutorial/inference/tensorflow/cifar10.py:18]: remove this
